tests/h/security/test_acl.py

```diff
@@ -0,0 +1,163 @@
+        )
+
+    def test_auth_client_with_matching_authority_may_add_members(
+        self, group, permitted_principals_for
+    ):
+        assert permitted_principals_for(Permission.Group.MEMBER_ADD) == {
+            f"client_authority:{group.authority}"
+        }
+
+    def test_auth_client_with_matching_authority_has_admin_permission(
+        self, group, permits
+    ):
+        assert permits(
+            ["noise", f"client_authority:{group.authority}"], Permission.Group.ADMIN
+        )
+        assert not permits(
+            ["noise", "client_authority:DIFFERENT_AUTHORITY"], Permission.Group.ADMIN
+        )
+
+    @pytest.mark.skip("This isn't currently true. Should it be?")
+    def test_admin_allowed_only_for_authority_when_no_creator(self, group, permits):
+        group.creator = None
+
+        assert permits(
+            ["noise", f"client_authority:{group.authority}"], Permission.Group.ADMIN
+        )
+
+    def test_staff_user_has_admin_permission_on_any_group(self, group, permits):
+        assert permits(["noise", role.Staff], Permission.Group.ADMIN)
+
+    def test_admin_user_has_admin_permission_on_any_group(self, group, permits):
+        assert permits(["noise", role.Admin], Permission.Group.ADMIN)
+
+    @pytest.mark.parametrize("readable_by", (ReadableBy.members, ReadableBy.world))
+    def test_creator_permissions(
+        self, group, permitted_principals_for, permits, readable_by
+    ):
+        group.readable_by = readable_by
+
+        assert permits(group.creator.userid, Permission.Group.ADMIN)
+        assert permits(group.creator.userid, Permission.Group.UPSERT)
+        assert permitted_principals_for(Permission.Group.MODERATE) == {
+            group.creator.userid
+        }
+
+    def test_no_creator_permissions_without_creator(
+        self, group, permitted_principals_for
+    ):
+        group.creator = None
+
+        assert not permitted_principals_for(Permission.Group.MODERATE)
+        assert not permitted_principals_for(Permission.Group.UPSERT)
+
+    def test_fallback_is_deny_all(self, group, permits):
+        assert not permits([security.Everyone], "non_existant_permission")
+
+    @pytest.fixture
+    def permits(self, group):
+        def permits(principals, permission):
+            class ACLCarrier:
+                def __acl__(self):
+                    return ACL.for_group(group)
+
+            return ACLAuthorizationPolicy().permits(
+                ACLCarrier(), principals, permission
+            )
+
+        return permits
+
+    @pytest.fixture
+    def permitted_principals_for(self, group):
+        def permitted_principals_for(permission):
+            class ACLCarrier:
+                def __acl__(self):
+                    return ACL.for_group(group)
+
+            return ACLAuthorizationPolicy().principals_allowed_by_permission(
+                ACLCarrier(), permission
+            )
+
+        return permitted_principals_for
```

Comment:
Eh some duplication here, but it's not the end of the world. With luck we're getting rid of this whole system soon.